File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.py
def solution(numbers, target):
    
    def dfs(depth, total):
        if depth == len(numbers):
            if total == target:
                return 1
            return 0
        
        return dfs(depth + 1, total + numbers[depth]) + dfs(depth + 1, total - numbers[depth])
    
    return dfs(0, 0)

File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    public int solution(int[] numbers, int target) {
        return dfs(numbers, target, 0, 0);
    }
    public static int dfs(final int[] numbers, final int target, int sum, int depth) {
        if (depth == numbers.length) {
            if (sum == target) {
                return 1;
            } else {
                return 0;
            }
        }

        return dfs(numbers, target, sum + numbers[depth], depth + 1)
                + dfs(numbers, target, sum - numbers[depth], depth + 1);
    }

}